<launch>
  <!-- Arguments -->
  <arg name="robot_namespace" default="engineer"/>
  <arg name="drive_version" default="v4"/>
  <arg name="map_file" default="$(find building_map)/maps/cleaned_map.yaml"/>
  <arg name="move_forward_only" default="false"/>

  <group ns="$(arg robot_namespace)">

  <!-- Engineer -->
  <!-- <arg name="multi_robot_name" default="$(arg robot_namespace)"/> -->
  <arg name="urdf_file" default="$(find xacro)/xacro --inorder '$(find engineer_description)/robots/engineer_default.urdf.xacro' robot_namespace:=$(arg robot_namespace) drive_version:=$(arg drive_version)"/>
  <param name="robot_description" command="$(arg urdf_file)"/>
  <!-- <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="50.0"/>
    <param name="tf_prefix" value="$(arg multi_robot_name)"/>
  </node> -->

	<!-- Map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)">
	<param name="frame_id" value="$(arg robot_namespace)/map"/>
	<remap from="map" to="/$(arg robot_namespace)/map"/>
	<remap from="map_metadata" to="/$(arg robot_namespace)/map_metadata"/>
  </node>

  <!-- AMCL -->
  <!-- Arguments -->
  <!-- <arg name="scan_topic"     default="$(arg robot_namespace)/laser_scan"/>
  <arg name="initial_pose_x" default="-3.0"/>
  <arg name="initial_pose_y" default="1.0"/>
  <arg name="initial_pose_a" default="0.0"/> -->

  <!-- AMCL -->
  <!-- <node pkg="amcl" type="amcl" name="amcl">

    <param name="use_map_topic"             value="true"/>
    <param name="min_particles"             value="500"/>
    <param name="max_particles"             value="3000"/>
    <param name="kld_err"                   value="0.02"/>
    <param name="update_min_d"              value="0.20"/>
    <param name="update_min_a"              value="0.20"/>
    <param name="resample_interval"         value="1"/>
    <param name="transform_tolerance"       value="0.5"/>
    <param name="recovery_alpha_slow"       value="0.00"/>
    <param name="recovery_alpha_fast"       value="0.00"/>
    <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
    <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
    <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
    <param name="gui_publish_rate"          value="50.0"/>

    <remap from="scan"                      to="/$(arg scan_topic)"/>
    <param name="laser_max_range"           value="3.5"/>
    <param name="laser_max_beams"           value="180"/>
    <param name="laser_z_hit"               value="0.5"/>
    <param name="laser_z_short"             value="0.05"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.5"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="laser_model_type"          value="likelihood_field"/>

    <param name="odom_model_type"           value="diff"/>
    <param name="odom_alpha1"               value="0.1"/>
    <param name="odom_alpha2"               value="0.1"/>
    <param name="odom_alpha3"               value="0.1"/>
    <param name="odom_alpha4"               value="0.1"/>
    <param name="odom_frame_id"             value="$(arg robot_namespace)/odom"/>
    <param name="base_frame_id"             value="$(arg robot_namespace)/base_footprint"/>

  </node> -->
  
  <!-- move_base -->
  <arg name="cmd_vel_topic" default="$(arg robot_namespace)/mobile_base_controller/cmd_vel" />
  <arg name="odom_topic" default="$(arg robot_namespace)/odom" />
  <!-- <arg name="base_global_planner" default="global_planner/GlobalPlanner"/> -->
  <arg name="base_global_planner" default="navfn/NavfnROS"/>

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="base_global_planner" value="$(arg base_global_planner)"/>
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    <!-- <rosparam file="$(find engineer_navigation)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find engineer_navigation)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find engineer_navigation)/config/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find engineer_navigation)/config/global_costmap_params.yaml" command="load" /> -->

    <!-- <rosparam file="$(find turtlebot3_navigation)/param/move_base_params.yaml" command="load" /> -->
	<!-- <rosparam file="$(find engineer_navigation)/config/base_local_planner_params.yaml" command="load"/> -->

    <!-- <rosparam file="$(find engineer_navigation)/config/local_planner_params.yaml" command="load" /> -->





	<rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_waffle.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_waffle.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find turtlebot3_navigation)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find turtlebot3_navigation)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find turtlebot3_navigation)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find turtlebot3_navigation)/param/dwa_local_planner_params_waffle.yaml" command="load" />





	<!-- global_costmap configuration -->
	<param name="global_costmap/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
	<param name="global_costmap/global_frame" value="$(arg robot_namespace)/map"/>
	<param name="global_costmap/robot_base_frame" value="$(arg robot_namespace)/base_footprint"/>
	<param name="global_costmap/scan/topic" value="/$(arg robot_namespace)/laser_scan"/>
	<param name="global_costmap/static_layer/trinary_costmap" value="false"/>
	<param name="global_costmap/scan/data_type" value="LaserScan"/>

	<!-- local_costmap configuration -->
	<param name="local_costmap/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
	<!-- <param name="local_costmap/global_frame" value="$(arg robot_namespace)/map"/> -->
	<param name="local_costmap/global_frame" value="$(arg robot_namespace)/odom"/>
	<param name="local_costmap/robot_base_frame" value="$(arg robot_namespace)/base_footprint"/>
	<param name="local_costmap/scan/topic" value="/$(arg robot_namespace)/laser_scan"/>
	<param name="local_costmap/scan/data_type" value="LaserScan"/>

    <remap from="cmd_vel" to="/$(arg cmd_vel_topic)"/>
    <remap from="odom" to="/$(arg odom_topic)"/>
    <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />
  </node>

  </group>

</launch>
